### PR TITLE
add chunked transfer to viewer

### PR DESCRIPTION
**Pull request by adameat**
### Changelog entry 

...

### Changelog category 

* New feature
* Not for changelog (changelog entry is not required)

### Additional information

adds asynchronous chunked transfer to viewer with two-way cancelling.
improves some aspects with http proxy and mon proxy.

### CODE REVIEW

File: ydb/core/blobstorage/ut_vdisk/lib/astest.h [hunk 1]
```diff
@@ -82,7 +82,6 @@ inline void TTestWithActorSystem::Run(NActors::IActor *testActor) {
         NKikimrServices::EServiceKikimr_Name
     );
     TString explanation;
-    logSettings->SetLevel(NLog::PRI_TRACE, NActorsServices::EServiceCommon::HTTP, explanation);
```
Comment (adameat) [inline ydb/core/blobstorage/ut_vdisk/lib/astest.h:85]: just something I mistakenly introduced in the previous commits

Comment (alexvru) [reply ydb/core/blobstorage/ut_vdisk/lib/astest.h:85]: don't need excuses

File: ydb/core/protos/mon.proto [hunk 2]
```diff
@@ -13,4 +13,11 @@ message TEvMonitoringRequest {
 
 message TEvMonitoringResponse {
     optional string HttpResponse = 1;
+    optional string DataChunk = 2;
```
Comment (alexvru) [inline ydb/core/protos/mon.proto:16]: really "string"? no non-utf8 payload possible?

File: ydb/core/viewer/json_pipe_req.cpp [hunk 3]
```diff
@@ -76,6 +76,61 @@ TViewerPipeClient::TViewerPipeClient(IViewer* viewer, NMon::TEvHttpInfo::TPtr& e
     }
 }
 
+TViewerPipeClient::TViewerPipeClient(IViewer* viewer, NHttp::TEvHttpProxy::TEvHttpIncomingRequest::TPtr& ev, const TString& handlerName)
+    : Viewer(viewer)
+    , HttpEvent(ev)
+{
+    TCgiParameters params(HttpEvent->Get()->Request->URL.After('?'));
+    NHttp::THeaders headers(HttpEvent->Get()->Request->Headers);
+    if (headers.Get("Content-Type") == "application/json") {
```
Comment (alexvru) [inline ydb/core/viewer/json_pipe_req.cpp:85]: and what if it is "application/json; encoding=utf-8" ?

File: ydb/core/viewer/json_pipe_req.h [hunk 4]
```diff
@@ -47,6 +47,7 @@ class TViewerPipeClient : public TActorBootstrapped<TViewerPipeClient> {
     NWilson::TSpan Span;
     IViewer* Viewer = nullptr;
     NMon::TEvHttpInfo::TPtr Event;
+    NHttp::TEvHttpProxy::TEvHttpIncomingRequest::TPtr HttpEvent;
```
Comment (alexvru) [inline ydb/core/viewer/json_pipe_req.h:50]: std::variant<> for this
or they can be both set at the same time?

File: ydb/core/viewer/viewer.cpp [hunk 5]
```diff
@@ -567,6 +585,56 @@ class TViewer : public TActorBootstrapped<TViewer>, public IViewer {
...
+        if (ev->Get()->Request->Method == "OPTIONS") {
+            TString uri(ev->Get()->Request->GetURI());
+            TString type = mimetypeByExt(uri.c_str());
+            if (type.empty()) {
+                type = "application/json";
+            }
+            TStringBuilder response;
+            response << "HTTP/1.1 204 No Content\r\n";
+            FillCORS(response, ev->Get());
+            response << "Content-Type: " + type + "\r\n"
```
Comment (alexvru) [inline ydb/core/viewer/viewer.cpp:599]: you really want to create temporary TString from "Content-Type: " + type + "\r\n" and then output it into stream and dispose after?

File: ydb/core/viewer/viewer.cpp [hunk 6]
```diff
@@ -567,6 +585,56 @@ class TViewer : public TActorBootstrapped<TViewer>, public IViewer {
...
+        TString path(ev->Get()->Request->GetURI());
+        auto itRedirect307 = Redirect307.find(path);
+        if (itRedirect307 != Redirect307.end()) {
+            TString redirect(ev->Get()->Request->URL);
+            redirect.erase(0, itRedirect307->first.size());
+            redirect.insert(0, itRedirect307->second);
+            TStringBuilder response;
+            response << "HTTP/1.1 307 Temporary Redirect\r\n";
+            FillCORS(response, ev->Get());
+            response << "Location: " + redirect + "\r\n"
```
Comment (alexvru) [inline ydb/core/viewer/viewer.cpp:613]: same here

File: ydb/core/viewer/viewer.h [hunk 7]
```diff
@@ -70,12 +73,68 @@ struct TRequestState {
...
+            return (*request)->Request.GetHeaders().HasHeader(name);
+        }
+        if (auto* request = std::get_if<const NHttp::TEvHttpProxy::TEvHttpIncomingRequest*>(&Request)) {
+            return NHttp::THeaders((*request)->Request->Headers).Has(name);
+        }
+        return false;
+    }
+
+    TString GetHeader(TStringBuf name) const {
+        if (auto* request = std::get_if<const NMon::TEvHttpInfo*>(&Request)) {
```
Comment (alexvru) [inline ydb/core/viewer/viewer.h:96]:
```
return std::visit(TOverloaded{
  [&](const NMon::TEvHttpInfo* ptr) { ... },
  [&](const ...) { ... }
}, Request);

```
would be much more readable and wouldn't create errors if one extends variant

Comment (adameat) [reply ydb/core/viewer/viewer.h:96]: readable? hm.

File: ydb/core/viewer/json_pipe_req.h (hunk 4 above)
Comment (adameat) [reply ydb/core/viewer/json_pipe_req.h:50]:
```
xenoxeno@mr-nvme-testing-01:/home-big/xenoxeno/work/ydb/ydb/core/viewer$ fgrep -rn "Event->" | wc -l
228
```

maybe later...

**Review by alexvru — DISMISSED** [review]

**Review by alexvru — APPROVED** [review]

**Pull request merged**